### PR TITLE
y can be a dict for numpy_input_fn

### PR DESCRIPTION
**Pull request by facaiy**
The PR is opened for #12610.

### What changes were proposed in this pull request?

dict is accept for `y` as labels.

### How was this patch tested?

+ [x] add unit test
+ [ ] pass all tests.

### CODE REVIEW

**Comment by martinwicke** [comment]
Marking as WIP until you have resolved test failures (it's not checked off the list yet, I assume it's not done).

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
Never mind, I'm running the tests.

**Comment by facaiy** [comment]
```python
//tensorflow/python/kernel_tests:matrix_solve_ls_op_test                TIMEOUT in 65.0s
  /var/lib/jenkins/workspace/tensorflow-pull-requests-cpu/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/workspace/bazel-out/local-opt/testlogs/tensorflow/python/kernel_tests/matrix_solve_ls_op_test/test.log

Executed 1205 out of 1205 tests: 1204 tests pass and 1 fails locally.
```

**Comment by facaiy** [comment]
I have only Linux GPU environment, hence the Linux CPU tests cannot be reproduced on my machine. Could anyone paste the test log? It is better to help me locate where is wrong. Thanks.

**Comment by facaiy** [comment]
Hi, could you retest the PR?

**Comment by drpngx** [comment]
Sure. Jenkins, test this please.

**Comment by drpngx** [comment]
@xiejw feel free to wait until it passes or review now.

**Comment by facaiy** [comment]
CI failed on MacOS CPU Tests. However, it cannot be reproduced on my mac.

```bash
Target //tensorflow/python:basic_session_run_hooks_test up-to-date:
  bazel-bin/tensorflow/python/basic_session_run_hooks_test
INFO: Elapsed time: 2881.694s, Critical Path: 152.93s
//tensorflow/python:basic_session_run_hooks_test                         PASSED in 17.9s

Executed 1 out of 1 test: 1 test passes.

(py27) ~/W/g/tensorflow ❯❯❯ python --version                            ENH/numpy_input_fn_support_dict ✱
Python 2.7.13 :: Anaconda 4.3.1 (x86_64)

(py27) ~/W/g/tensorflow ❯❯❯ uname -a                                    ENH/numpy_input_fn_support_dict ✱
Darwin bj-m-203050a.local 15.6.0 Darwin Kernel Version 15.6.0: Tue Apr 11 16:00:51 PDT 2017; root:xnu-3248.60.11.5.3~1/RELEASE_X86_64 x86_64
```

**Comment by facaiy** [comment]
I cannot reproduce the failure on mac test:

```bash
(py27) ~/W/g/tensorflow ❯❯❯ bazel test -c opt //tensorflow/python:basic_session_run_hooks_test
WARNING: /Users/facai/Workshop/github/tensorflow/tensorflow/core/BUILD:1634:1: in includes attribute of cc_library rule //tensorflow/core:framework_headers_lib: '../../external/nsync/public' resolves to 'external/nsync/public' not below the relative path of its package 'tensorflow/core'. This will be an error in the future. Since this rule was created by the macro 'cc_header_only_library', the error might have been caused by the macro implementation in /Users/facai/Workshop/github/tensorflow/tensorflow/tensorflow.bzl:911:30.
INFO: Found 1 test target...
Target //tensorflow/python:basic_session_run_hooks_test up-to-date:
  bazel-bin/tensorflow/python/basic_session_run_hooks_test
INFO: Elapsed time: 19.859s, Critical Path: 17.47s
//tensorflow/python:basic_session_run_hooks_test                         PASSED in 17.5s

Executed 1 out of 1 test: 1 test passes.
```

**Comment by martinwicke** [comment]
This test failure is unrelated (and flaky, I believe).

File: tensorflow/python/estimator/inputs/numpy_io_test.py [hunk 1]
```diff
@@ -285,6 +285,51 @@ def testNumpyInputFnWithMismatchLengthOfInputs(self):
...
+      self.assertAllEqual(targets['y1'], [-32, -31])
+      self.assertAllEqual(targets['y2'], [32, 31])
+
+      session.run([features_tensor, targets_tensor])
+      with self.assertRaises(errors.OutOfRangeError):
+        session.run([features_tensor, targets_tensor])
+
+      coord.request_stop()
+      coord.join(threads)
+
```
Comment (martinwicke) [inline tensorflow/python/estimator/inputs/numpy_io_test.py]: Please add a test for case `y={}`. Ideally it would simply work, but without a test, who knows?

**Review by martinwicke — CHANGES_REQUESTED** [review]
Looks good otherwise.

File: tensorflow/python/estimator/inputs/numpy_io_test.py (hunk 1 above)
Comment (facaiy) [reply tensorflow/python/estimator/inputs/numpy_io_test.py]: Thank you for review, @martinwicke .  a `y={}` test case has been added below.

File: tensorflow/python/estimator/inputs/numpy_io_test.py [hunk 2]
```diff
@@ -285,6 +285,77 @@ def testNumpyInputFnWithMismatchLengthOfInputs(self):
...
+  def testNumpyInputFnWhenLabelIsEmptyDictionary(self):
+    a = np.arange(4) * 1.0
+    b = np.arange(32, 36)
+    x = {'a': a, 'b': b}
+    y = {}
+
+    with self.test_session() as session:
+      input_fn = numpy_io.numpy_input_fn(
+        x, y, batch_size=2, shuffle=False, num_epochs=1)
+      features_tensor = input_fn()
```
Comment (martinwicke) [inline tensorflow/python/estimator/inputs/numpy_io_test.py]: Is this the right behavior? Or would we expect the input function generated to return an empty labels dict?

Comment (facaiy) [reply tensorflow/python/estimator/inputs/numpy_io_test.py]: I have no idea. `y={}` seems a little weird. Do you prefer to returning a empty labels dict?

Comment (facaiy) [reply tensorflow/python/estimator/inputs/numpy_io_test.py]: By the way,  is `x = {}` or `y = {}` a valid argument?

Comment (martinwicke) [reply tensorflow/python/estimator/inputs/numpy_io_test.py]: `x = {}` makes no sense (testing for it and raise an appropriate error would be nice), `y = {}` does make sense for unlabeled problems. As does `y=None`.

However, I got confused because even though the docs don't mention it, this function sometimes returns only `features`, and sometimes it returns `(features, labels)`. 

I would find it somewhat odd if the return type changed based on the content of the dict (or non-content).

Comment (facaiy) [reply tensorflow/python/estimator/inputs/numpy_io_test.py]: I agree that we'd better return consistent result `(features, labels)`.  How about this:
+ When x = {}, raise an ValueError
+ When y = None, labels = None
+ When y = {}, labels = {}

Comment (martinwicke) [reply tensorflow/python/estimator/inputs/numpy_io_test.py]: That's perfect.

Comment (facaiy) [reply tensorflow/python/estimator/inputs/numpy_io_test.py]: Oh, I forget to remind you that the compatibility will be broken if we return `None` when `y=None`.

**Comment by facaiy** [comment]
Hi, I revise code and explain the invalid cases as below:

+ `x = {}`, raise a ValueError
+ `y = {}`, raise a ValueError, and suggest user to use `None` instead.
+ `y = None`, leave a todo comment, but don't change the behavior now for compatibility.

By the way, more tests have been added to cover those cases above.

**Comment by facaiy** [comment]
Moreover, `shuffle` seems an optional argument (default value is `None`), but it requires a boolean value. 

```python
def numpy_input_fn(x,
                    y=None,
                    batch_size=128,
                    num_epochs=1,
                    shuffle=None,
                    queue_capacity=1000,
                    num_threads=1):
```

Really odd, if `shuffle` is required, why to give a default value?  If it is optional, why to give a invalid value by default?

**Comment by facaiy** [comment]
Hi, quite a long time, might we test the PR? Thanks.

**Comment by CSNoyes** [comment]
Is this going to be merged?

**Comment by drpngx** [comment]
@facaiy did you address @martinwicke comments?

**Comment by facaiy** [comment]
I think so, 

> Hi, I revise code and explain the invalid cases as below:
> x = {}, raise a ValueError
> y = {}, raise a ValueError, and suggest user to use None instead.
> y = None, leave a todo comment, but don't change the behavior now for compatibility.
> By the way, more tests have been added to cover those cases above.

**Comment by facaiy** [comment]
Hi, @martinwicke @xiejw . What do you think about the behavior changed ? Thanks.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Review by martinwicke — APPROVED** [review]
